utxoscanner.go

```diff
@@ -0,0 +1,517 @@
+		if !fetch {
+			options := rescanOptions{
+				watchList: filterEntries,
+			}
+
+			match, err := s.cfg.BlockFilterMatches(&options, hash)
+			if err != nil {
+				return spillRequests(), err
+			}
+
+			// If still no match is found, we have no reason to
+			// fetch this block, and can continue to next height.
+			if !match {
+				continue
+			}
+		}
+
+		log.Debugf("Fetching block height=%d hash=%s", height, hash)
+
+		block, err := s.cfg.GetBlock(*hash)
+		if err != nil {
+			return spillRequests(), err
+		}
+
+		// Check to see if the utxoscanner has been signaled to exit.
+		select {
+		case <-s.quit:
+			return spillRequests(), ErrShuttingDown
+		default:
+		}
+
+		log.Debugf("Got block height=%d hash=%s", height, hash)
+
+		// Grab the txns that created the outputs requested at this
+		// block height.
+		txnsInBlock := findInitialTransactions(block.MsgBlock(), reqs)
+		for outPoint, tx := range txnsInBlock {
+			if tx != nil {
+				log.Debugf("Block %d creates output %s",
+					height, outPoint)
+			}
+			initialTx[outPoint] = tx
+		}
+
+		// Filter the block for any spends using our set of outpoints.
+		spends := findSpends(block.MsgBlock(), height, outpoints)
+		for outPoint, spend := range spends {
+			// Find the request this spend relates to.
+			request, ok := requests[outPoint]
+			if !ok {
+				log.Debugf("Utxo scan request for scanned "+
+					"outpoint=%v was not found", outPoint)
+				continue
+			}
+
+			// With the request located, we remove this outpoint
+			// from both the requests, outpoints, and initial txns
+			// map. This will ensures we don't continue watching
+			// this outpoint.
+			delete(requests, outPoint)
+			delete(outpoints, outPoint)
+			delete(initialTx, outPoint)
+
+			log.Debugf("Outpoint %s is spent in tx %s",
+				outPoint.String(), spew.Sprint(spend))
+
+			// Deliver the spend report to the request.
+			request.Result(spend, nil)
+		}
+
+		// Rebuild filter entries from cached entries remaining in
+		// outpoints map.
+		filterEntries = filterEntries[:0]
+		for _, entry := range outpoints {
+			filterEntries = append(filterEntries, entry)
+		}
+	}
+
+	// We've scanned up to the end height, now perform a check to see if we
+	// still have any new blocks to process. If this is the first time
+	// through, we might have a few blocks that were added since the
+	// scan started.
+	currStamp, err := s.cfg.BestSnapshot()
+	if err != nil {
+		return spillRequests(), err
+	}
+
+	// If the returned height is higher, we still have more blocks to go.
+	// Shift the start and end heights and continue scanning.
+	if uint32(currStamp.Height) > endHeight {
+		startHeight = endHeight + 1
+		endHeight = uint32(currStamp.Height)
+		goto scanToEnd
+	}
+
+	log.Debugf("Finished batch, %d unspent outpoints", len(requests))
+
+	for _, request := range requests {
+		tx, ok := initialTx[*request.OutPoint]
+		if ok {
+			request.Result(tx, nil)
+		} else {
+			// A nil SpendReport indicates the output was not found.
```

Comment:
It seems there isn't a distinction between the output being not found, and not _spent_.

Reply:
The above clause handles the case where the output is not spent, as it will return the initial txn. The above clause can also pass nil `tx`.